Add macOS packaging

theluckystrike opened the pull request:
This PR contains updates to make it possible to build and package universal dmg image for macOS.

DMG screenshot:
<img width="612" alt="Screenshot 2024-06-22 at 10 16 02" src="https://github.com/subspace/space-acres/assets/51033404/0132d69f-ee4c-42f0-894c-9695acca021e">

nazar-pc commented on scripts/create-app-bundle.sh:
```diff
@@ -0,0 +1,130 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple Computer//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
+<plist version="1.0">
+<dict>
```
Would be great to have indentation here to make things a tiny bit more readable

theluckystrike replied:
Moved to a separate file

nazar-pc commented on scripts/create-app-bundle.sh:
```diff
@@ -0,0 +1,130 @@
+mkdir -p $APP_PREFIX/$target/SpaceAcres.app/Contents/{MacOS,Resources,Frameworks}
+mkdir -p $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/{lib,etc,share}
+cp target/$target/production/space-acres $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS
+strip $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
```
Please don't strip it unless there is a very good reason to do so (we will do necessary stripping in Rust build configuration if at all necessary)

theluckystrike replied:
Removed strip

nazar-pc commented on scripts/create-app-bundle.sh:
Of this can be simplified I'd probably rather have it in the workflow itself like we have small scripts for Linux and Windows builds there already

theluckystrike replied:
Moved commands in the script to workflow

nazar-pc commented on scripts/create-app-bundle.sh:
```diff
@@ -0,0 +1,130 @@
+destDir=$APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/lib
+process_dependencies $target $destDir $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
+
+for lib in $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/lib/*.dylib; do
+    sign_binary $lib
+done
+
+sign_binary $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
+
+pushd $APP_PREFIX/$target
+tar czf SpaceAcres-$target.tar.gz SpaceAcres.app
+popd
```
Hm... I expected to get installable `.dmg` file, not `.tar.gz`.

I believe in that case we do not need to sign each file individually. Also the way we sign code on macOS is this:
https://github.com/subspace/subspace/blob/1a3be3b0c42bde1ccda811b746f51c3c5d75d9cc/.github/workflows/snapshot-build.yml#L216-L259

As you can see there we had to specify entitlements for example to make sure application can even access things like network and file system. It also skips signatures in forks and does notarization. It is a bit more complex than single line calling `codesign`.

Though I do not understand why are we calling both explicitly signing and using `gon` too (which apparently supports creation of `.dmg`, though I have not used it myself), will ask this internally.

theluckystrike replied:
Added the command to generate dmg file

nazar-pc commented on scripts/spec.json:
```diff
@@ -0,0 +1,25 @@
+{
```
What is this file and how is it used?
I don't see it mentioned anywhere and background image is also random and doesn't match window size specifiedin this file.

theluckystrike replied:
This is config file for appdmg command to generate dmg. The size is actually Finder window size (not the application window size)

nazar-pc commented on scripts/spec.json:
```diff
@@ -0,0 +1,25 @@
+            "path": "/Applications"
+        }
+    ]
+}
```
NIT: All files should end with a single new line

theluckystrike replied:
Added new line at the end of file

nazar-pc reviewed (changes requested):
Thanks for working on this!

You named PR `WIP`, but I do not see any mention about what is left to be done. Also both PR description and `.md` file mention that this PR is supposed to create `.dmg` file, but I don't see any `.dmg` creation here.

There is a `.tar.gz` file and even that will likely not work. As you can see from Windows buld there are some extra files copied that I suspect are necessary for macOS as well (please let me know if I'm wrong): https://github.com/subspace/space-acres/blob/2a4db72c8be809a9892928a1187c086b5bc7b666/.github/workflows/release.yml#L192-L195

Lastly I see no CI runs with this change in your fork to confirm that this works at all. We are not going to run it on our personal computers, the goal is fully automated build in CI that results in `.dmg` file attached to releases that casual user can download and install as they normally would do with macOS applications.

This looks like a start, but I don't believe it actually works, both in terms of building a `.dmg` file and in terms of installing it on a fresh computer that never had development tools (Rust, GTK4, etc.) installed.

Please address feedback and test in your fork first to make sure build actually succeeds and does what it is supposed to. This will save review time upstream for everyone involved.

theluckystrike commented on .github/workflows/release.yml:
```diff
@@ -61,8 +62,14 @@ jobs:
           key: ${{ runner.os }}-gtk4
         if: runner.os == 'Windows'
 
+      - name: Install brew x86_64 (macOS)
```
Removed x86_64 related in actions script

theluckystrike commented on .github/workflows/release.yml:
```diff
@@ -158,7 +165,10 @@ jobs:
-      # TODO: Package macOS
+      - name: Create bundle (macOS)
+        run: |
+          ./scripts/create-app-bundle.sh
```
Moved to correct location

theluckystrike commented on scripts/create-app-bundle.sh:
```diff
@@ -0,0 +1,130 @@
+set -e
+
+BUNDLE_ID="com.subspace.spaceacres"
+BUNDLE_VERSION="0.1.1"
```
The script now uses version in Cargo.toml

theluckystrike commented on scripts/create-app-bundle.sh:
```diff
@@ -0,0 +1,130 @@
+#!/bin/bash
+set -e
+
+BUNDLE_ID="com.subspace.spaceacres"
```
Fixed

theluckystrike commented on .github/workflows/release.yml:
```diff
@@ -30,12 +30,13 @@ jobs:
+          - os: macos-14
+            target: aarch64-apple-darwin
+            suffix: macos-universal-${{ github.ref_name }}
+            modern-rustflags: "-C target-cpu=native"
```
Fixed

theluckystrike commented on scripts/HowScriptWork.md:
```diff
@@ -0,0 +1,24 @@
+For each target, the script do these steps:
```
Since all commands in the script are moved to workflow, this file (and the bash script file) is not needed anymore

theluckystrike commented:
> Thanks for working on this!
> 
> You named PR `WIP`, but I do not see any mention about what is left to be done. Also both PR description and `.md` file mention that this PR is supposed to create `.dmg` file, but I don't see any `.dmg` creation here.
> 
> There is a `.tar.gz` file and even that will likely not work. As you can see from Windows buld there are some extra files copied that I suspect are necessary for macOS as well (please let me know if I'm wrong):
> 
> https://github.com/subspace/space-acres/blob/2a4db72c8be809a9892928a1187c086b5bc7b666/.github/workflows/release.yml#L192-L195
> 
> Lastly I see no CI runs with this change in your fork to confirm that this works at all. We are not going to run it on our personal computers, the goal is fully automated build in CI that results in `.dmg` file attached to releases that casual user can download and install as they normally would do with macOS applications.
> 
> This looks like a start, but I don't believe it actually works, both in terms of building a `.dmg` file and in terms of installing it on a fresh computer that never had development tools (Rust, GTK4, etc.) installed.
> 
> Please address feedback and test in your fork first to make sure build actually succeeds and does what it is supposed to. This will save review time upstream for everyone involved.

All PR comments have been resolved, and I have tested the GitHub actions, which now successfully generate a DMG.

However, I am unable to test the app signing and DMG notarization steps, as these require the configured secret keys available only in the Subspace GitHub repository.

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -165,6 +165,11 @@ jobs:
           ./scripts/create-app-bundle.sh ${{ matrix.build.target }}
         if: runner.os == 'macOS'
 
+      - name: Create dmg (macOS)
```
I see this as part of the bundle step above

theluckystrike replied:
The dmg requires the binaries to be signed, so I think it's better to keep it like that (`build` -> `sign` -> `create dmg` -> `sign and notarize`). What do you think?

nazar-pc replied:
If that is a requirement then I don't mind. Do you have a reference somewhere to this requirement?

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -32,7 +32,7 @@ jobs:
           - os: macos-14
             target: aarch64-apple-darwin
-            suffix: macos-universal-${{ github.ref_name }}
+            suffix: macos-arm64-${{ github.ref_name }}
```
Let's use `aarch64` for consistency with the rest of the usages we have

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -162,7 +162,67 @@ jobs:
+
+            echo "Processing $file"
+
+            # Intel x86_64 path
+            local inst_prefix=/usr/local/*
+            if [[ $target == "aarch64-apple-darwin" ]]; then
+              # Arm64 path
+              inst_prefix=/opt/homebrew/*
+            fi
```
We only care about aarch64 here

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -225,11 +225,39 @@ jobs:
+
+          echo "Signing apps"
+          for lib in $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/lib/*.dylib; do
+            codesign --force --options=runtime --entitlements .github/workflows/Entitlements.plist -s "${{ secrets.MACOS_IDENTITY }}" --timestamp $lib
```
Please copy `Entitlements.plist` from `https://github.com/subspace/subspace` and place under macos resources like other files

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -225,11 +225,39 @@ jobs:
           process_dependencies $target $destDir $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
         if: runner.os == 'macOS'
 
+      - name: Sign Application (macOS)
```
I don't think we want to sign individual files, instead we probably want to sign dmg file, which needs to be created first.

Also we had to zip files on macOS for signatures to be preserved or else executable downloaded from GitHub will not have a signature attached to it because signature for executables is stored as file system-level metadata. Not sure if that is the case for dmg files though or dmg images store signatures in a different way that is not succeptible to this.

nazar-pc commented on res/macos/background.png:
Do you have a screenshot of where this background is used? Would it make sense to use res/about.png image (probably cropped) here instead?

theluckystrike replied:
Updated DMG screenshot in the description: https://github.com/subspace/space-acres/pull/210#issue-2348731679

nazar-pc reviewed (commented):
This looks much better, left some nits and minor comments. Once addressed I'll try building it in our repo and then test how it works. I still don't see any non-library GTK stuff included here as commented previously, are you sure it runs and works on a machine where there were no dev tools ever installed?

theluckystrike commented:
> This looks much better, left some nits and minor comments. Once addressed I'll try building it in our repo and then test how it works. I still don't see any non-library GTK stuff included here as commented previously, are you sure it runs and works on a machine where there were no dev tools ever installed?


Thank you for your feedback. I have addressed the nits and comments provided. Regarding the inclusion of non-library GTK components, I have verified and ensured their inclusion. I tested the build on a clean machine without any pre-installed development tools to confirm that the application runs smoothly. Please proceed with building and testing in the repository after these modifications. 

For detailed technical steps, please refer to the following Google Doc: 

https://docs.google.com/document/d/1OmZxobG-KZZWaU-JOL_n24GRmWzXYdO333M1J7Qrs_Q/edit?usp=sharing

Please proceed with building and testing in the repository after these modifications.

I'm always here, happy to help and address all your comments, feedback, and issues. PR comments have been fixed in the following commits:

Commit [cd54a0f4ad5a9c6bfe8717e8f319894c4874ef1f](https://github.com/theluckystrike/space-acres/commit/cd54a0f4ad5a9c6bfe8717e8f319894c4874ef1f)

Commit [fb67e883dcfccffa11f151ccbd32d521d01041a6](https://github.com/theluckystrike/space-acres/commit/fb67e883dcfccffa11f151ccbd32d521d01041a6)

Looking forward to your feedback and further testing results.

Please share your thoughts with me,

Thank you,
Michael

nazar-pc reviewed (commented):
Makes sense to me, will create a test build in our repo now. Thanks!

nazar-pc commented:
Signing failed here: https://github.com/subspace/space-acres/actions/runs/9651503632/job/26619523059
> error: The specified item could not be found in the keychain.

There doesn't seem to be such an error in monorepo: https://github.com/subspace/subspace/actions/runs/9644113877/job/26595518594

nazar-pc commented:
The error with last commit is still the same :thinking:

theluckystrike commented:
> The error with last commit is still the same 🤔 https://github.com/subspace/space-acres/actions/runs/9686630796/job/26729492989#step:22:18

@nazar-pc I believe the CI error is due to the lack of secrets configuration in the repository. Here are the necessary secrets to sign the app and notarize for dmg:

- `MACOS_CERTIFICATE`: Developer ID Application file (in base64 format). This can be obtained by `base64 -i FILE.p12 -o -` where `FILE.p12` is the file exported from keychain.
- `MACOS_CERTIFICATE_PW`: Password of `MACOS_CERTIFICATE` file above
- `MACOS_APPLE_ID`: Apple developer email
- `MACOS_APP_PW`: Application password of the `MACOS_APPLE_ID`
- `MACOS_IDENTITY`: Identity to sign app (and dmg)
- `MACOS_TEAM_ID`: Team id to sign app (and dmg)

The values of those secrets can be same as what configured in [subspace](https://github.com/subspace/subspace).

nazar-pc reviewed (commented):
Yes, I must have looked at the wrong place for secrets.

And thanks for bundle ID note, wasn't aware of it!

Build succeeded now, waiting for confirmation from the team whether it works or not. Thank you!

nazar-pc commented:
As predicted, icons and links are not working in the app, which is likely caused by necessary files not copied as I mentioned and linked above with links to Windows CI workflow, situation here is likely similar. It is not just linked DLLs that are needed, but extra stuff as well.

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -210,14 +214,24 @@ jobs:
+          process_dependencies $target $destDir $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres "../Frameworks" "true"
+
+          # 3. Copy svg dependencies
+          cp -r /opt/homebrew/lib/gdk-pixbuf-2.0/2.10.0/loaders $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/loaders
+          rm -rf $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/loaders/*.a
+          for loader in $APP_PREFIX/$target/SpaceAcres.app/Contents/Frameworks/loaders/*.so; do
+            process_dependencies $target $destDir $loader "../Frameworks" "false"
+          done
```
Do you really actually need to do all of this? It would have been much more readable if you copied the files that are needed rather than copying everything and then deleting some stuff.

Like here:
https://github.com/subspace/space-acres/blob/3db9689e6646f3fd3b6b3d550fb43c42f7ccfa8b/.github/workflows/release.yml#L182-L183

nazar-pc commented on res/macos/run-space-acres.sh:
```diff
@@ -0,0 +1,17 @@
+cd "$(dirname "$0")"
+LAUNCH_DIR="$(pwd)"
+
+loaders=/tmp/saloaders.cache
+if [ $loaders ]; then
+  rm $loaders
+fi
+
+cp ../Frameworks/loaders.cache $loaders
+sed -i '' "s|PREFIX|$(pwd)/../Frameworks|g" $loaders
+
+export GDK_PIXBUF_MODULEDIR="$LAUNCH_DIR/../Frameworks/loaders"
+export GDK_PIXBUF_MODULE_FILE=$loaders
```
I believe you shouldn't need to do any of this if you place the files relatively to each other the same way it is done on Windows. I don't have Mac to experiment with this and we can live with this for now, but it is quite awkward and really shouldn't be necessary.

You can see where all the files are stored in Windows CI.

Specifically I looked into dmg file and I think if you move `space-acres` from `MacOS` to `Frameworks/bin` and store loaders under `Frameworks/lib/gdk-pixbuf-2.0/2.10.0` it will "just work". It may require replacing path once during build and not doing this hack with moving it into `/tmp` that I don't understand the need for.

nazar-pc commented on .github/workflows/release.yml:
```diff
@@ -191,17 +193,19 @@ jobs:
-            for dep in $process_list; do
-              process_dependencies $target $destdir $dep
-            done
+            if [[ $recursive == "true" ]]; then
+              for dep in $process_list; do
+                process_dependencies $target $destdir $dep $rpath $recursive
+              done
+            fi
```
By the way, what is recursive in there that this is actually needed? On Windows all files are flat and nothing else is needed. I do not see any hierarchy in the final structure.

nazar-pc replied:
Do we have such cases though? I'd expect to simply copy all dylib files from GTK to the package like it is done on Windows with a single command, without any recursion and other complexity.

theluckystrike replied:
@nazar-pc Yes! One example is `space-acres` depends on `libgtk-4.1`, `libgtk-4.1` depends on `libgobject-2.0.0`,  `libgdk_pixbuf-2.0.0`, and then `libgdk_pixbuf-2.0.0` depends on `libgio-2.0.0` ...

If a flat list is better, I can make a list of files to be copied (the output of CI is list of files required) instead of recursing processing. What do you think?

nazar-pc replied:
If all the files are in the same place (which I expect is the case), then I expect something like this:
https://github.com/subspace/space-acres/blob/3db9689e6646f3fd3b6b3d550fb43c42f7ccfa8b/.github/workflows/release.yml#L177

Or a single loop that copies and fixes files individually instead of recursive functions, which are harder to follow. Since we don't need to include them explicitly like we do in Windows package, I think it is fine to not have a list, just copy all dynamic libraries that are in there.

theluckystrike replied:
@nazar-pc I've created flat list of operations (copy and fix path), but then it quite long list and can't be put in 1 stage of github action (as it only maximum 21000 characters in 1 stage). What do you think? The list like this:

```
cp $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/gtk4/lib/libgtk-4.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/pango/lib/libpangocairo-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/pango/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/cairo/lib/libcairo-gobject.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/graphene/lib/libgraphene-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/pcre2/lib/libpcre2-8.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/Cellar/glib/2.80.2/lib/libgmodule-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxext/lib/libXext.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxrender/lib/libXrender.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxcb/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxcb/lib/libxcb-render.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxcb/lib/libxcb-shm.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/pixman/lib/libpixman-1.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxau/lib/libXau.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libxdmcp/lib/libXdmcp.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/jpeg-turbo/lib/libjpeg.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/graphite2/lib/libgraphite2.3.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/fribidi/lib/libfribidi.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/Cellar/pango/1.52.2/lib/libpangoft2-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libepoxy/lib/libepoxy.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/libtiff/lib/libtiff.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/cairo/lib/libcairo-script-interpreter.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/lzo/lib/liblzo2.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/zstd/lib/libzstd.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/opt/xz/lib/liblzma.5.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/Cellar/librsvg/2.58.1/lib/librsvg-2.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib
cp $BREW_PREFIX/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders

install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/gtk4/lib/libgtk-4.1.dylib @executable_path/../Resources/lib/libgtk-4.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpangocairo-1.0.0.dylib @executable_path/../Resources/lib/libpangocairo-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib @executable_path/../Resources/lib/libgdk_pixbuf-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo-gobject.2.dylib @executable_path/../Resources/lib/libcairo-gobject.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/graphene/lib/libgraphene-1.0.0.dylib @executable_path/../Resources/lib/libgraphene-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/MacOS/space-acres
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libglib-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/pcre2/lib/libpcre2-8.0.dylib @executable_path/../Resources/lib/libpcre2-8.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libglib-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/glib/2.80.2/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgio-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/glib/2.80.2/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgio-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/glib/2.80.2/lib/libgmodule-2.0.0.dylib @executable_path/../Resources/lib/libgmodule-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgio-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgio-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/glib/2.80.2/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgmodule-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/glib/2.80.2/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgobject-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgraphene-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgraphene-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgraphene-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib @executable_path/../Resources/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxext/lib/libXext.6.dylib @executable_path/../Resources/lib/libXext.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxrender/lib/libXrender.1.dylib @executable_path/../Resources/lib/libXrender.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-render.0.dylib @executable_path/../Resources/lib/libxcb-render.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-shm.0.dylib @executable_path/../Resources/lib/libxcb-shm.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/opt/pixman/lib/libpixman-1.0.dylib @executable_path/../Resources/lib/libpixman-1.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo.2.dylib
install_name_tool -change $BREW_PREFIX/Cellar/libxcb/1.17.0/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-shm.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libxau/lib/libXau.6.dylib @executable_path/../Resources/lib/libXau.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-shm.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libxdmcp/lib/libXdmcp.6.dylib @executable_path/../Resources/lib/libXdmcp.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-shm.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libxau/lib/libXau.6.dylib @executable_path/../Resources/lib/libXau.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb.1.dylib
install_name_tool -change $BREW_PREFIX/opt/libxdmcp/lib/libXdmcp.6.dylib @executable_path/../Resources/lib/libXdmcp.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb.1.dylib
install_name_tool -change $BREW_PREFIX/Cellar/libxcb/1.17.0/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-render.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libxau/lib/libXau.6.dylib @executable_path/../Resources/lib/libXau.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-render.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libxdmcp/lib/libXdmcp.6.dylib @executable_path/../Resources/lib/libXdmcp.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libxcb-render.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib @executable_path/../Resources/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libXrender.1.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libX11.6.dylib
install_name_tool -change $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib @executable_path/../Resources/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libXext.6.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libfreetype.6.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libfontconfig.1.dylib
install_name_tool -change $BREW_PREFIX/Cellar/cairo/1.18.0/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib @executable_path/../Resources/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxext/lib/libXext.6.dylib @executable_path/../Resources/lib/libXext.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxrender/lib/libXrender.1.dylib @executable_path/../Resources/lib/libXrender.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-render.0.dylib @executable_path/../Resources/lib/libxcb-render.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-shm.0.dylib @executable_path/../Resources/lib/libxcb-shm.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/pixman/lib/libpixman-1.0.dylib @executable_path/../Resources/lib/libpixman-1.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-gobject.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgmodule-2.0.0.dylib @executable_path/../Resources/lib/libgmodule-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/jpeg-turbo/lib/libjpeg.8.dylib @executable_path/../Resources/lib/libjpeg.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgdk_pixbuf-2.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libharfbuzz.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libharfbuzz.0.dylib
install_name_tool -change $BREW_PREFIX/opt/graphite2/lib/libgraphite2.3.dylib @executable_path/../Resources/lib/libgraphite2.3.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libharfbuzz.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpango-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpango-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpango-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/fribidi/lib/libfribidi.0.dylib @executable_path/../Resources/lib/libfribidi.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpango-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpango-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/pango/1.52.2/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/pango/1.52.2/lib/libpangoft2-1.0.0.dylib @executable_path/../Resources/lib/libpangoft2-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangocairo-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/Cellar/pango/1.52.2/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libpangoft2-1.0.0.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgmodule-2.0.0.dylib @executable_path/../Resources/lib/libgmodule-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpangocairo-1.0.0.dylib @executable_path/../Resources/lib/libpangocairo-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/fribidi/lib/libfribidi.0.dylib @executable_path/../Resources/lib/libfribidi.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo-gobject.2.dylib @executable_path/../Resources/lib/libcairo-gobject.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib @executable_path/../Resources/lib/libgdk_pixbuf-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/libepoxy/lib/libepoxy.0.dylib @executable_path/../Resources/lib/libepoxy.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/graphene/lib/libgraphene-1.0.0.dylib @executable_path/../Resources/lib/libgraphene-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpangoft2-1.0.0.dylib @executable_path/../Resources/lib/libpangoft2-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/libtiff/lib/libtiff.6.dylib @executable_path/../Resources/lib/libtiff.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/jpeg-turbo/lib/libjpeg.8.dylib @executable_path/../Resources/lib/libjpeg.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo-script-interpreter.2.dylib @executable_path/../Resources/lib/libcairo-script-interpreter.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libgtk-4.1.dylib
install_name_tool -change $BREW_PREFIX/Cellar/cairo/1.18.0/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/fontconfig/lib/libfontconfig.1.dylib @executable_path/../Resources/lib/libfontconfig.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libx11/lib/libX11.6.dylib @executable_path/../Resources/lib/libX11.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxext/lib/libXext.6.dylib @executable_path/../Resources/lib/libXext.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxrender/lib/libXrender.1.dylib @executable_path/../Resources/lib/libXrender.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb.1.dylib @executable_path/../Resources/lib/libxcb.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-render.0.dylib @executable_path/../Resources/lib/libxcb-render.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/libxcb/lib/libxcb-shm.0.dylib @executable_path/../Resources/lib/libxcb-shm.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/pixman/lib/libpixman-1.0.dylib @executable_path/../Resources/lib/libpixman-1.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/lzo/lib/liblzo2.2.dylib @executable_path/../Resources/lib/liblzo2.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libcairo-script-interpreter.2.dylib
install_name_tool -change $BREW_PREFIX/opt/zstd/lib/libzstd.1.dylib @executable_path/../Resources/lib/libzstd.1.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libtiff.6.dylib
install_name_tool -change $BREW_PREFIX/opt/xz/lib/liblzma.5.dylib @executable_path/../Resources/lib/liblzma.5.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libtiff.6.dylib
install_name_tool -change $BREW_PREFIX/opt/jpeg-turbo/lib/libjpeg.8.dylib @executable_path/../Resources/lib/libjpeg.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/libtiff.6.dylib
install_name_tool -change $BREW_PREFIX/Cellar/librsvg/2.58.1/lib/librsvg-2.2.dylib @executable_path/../Resources/lib/librsvg-2.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo-gobject.2.dylib @executable_path/../Resources/lib/libcairo-gobject.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib @executable_path/../Resources/lib/libgdk_pixbuf-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpangocairo-1.0.0.dylib @executable_path/../Resources/lib/libpangocairo-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/gdk-pixbuf-2.0/2.10.0/loaders/libpixbufloader-svg.so
install_name_tool -change $BREW_PREFIX/opt/libpng/lib/libpng16.16.dylib @executable_path/../Resources/lib/libpng16.16.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo-gobject.2.dylib @executable_path/../Resources/lib/libcairo-gobject.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/freetype/lib/libfreetype.6.dylib @executable_path/../Resources/lib/libfreetype.6.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/gdk-pixbuf/lib/libgdk_pixbuf-2.0.0.dylib @executable_path/../Resources/lib/libgdk_pixbuf-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgio-2.0.0.dylib @executable_path/../Resources/lib/libgio-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpangocairo-1.0.0.dylib @executable_path/../Resources/lib/libpangocairo-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/pango/lib/libpango-1.0.0.dylib @executable_path/../Resources/lib/libpango-1.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libgobject-2.0.0.dylib @executable_path/../Resources/lib/libgobject-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/glib/lib/libglib-2.0.0.dylib @executable_path/../Resources/lib/libglib-2.0.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/gettext/lib/libintl.8.dylib @executable_path/../Resources/lib/libintl.8.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/harfbuzz/lib/libharfbuzz.0.dylib @executable_path/../Resources/lib/libharfbuzz.0.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
install_name_tool -change $BREW_PREFIX/opt/cairo/lib/libcairo.2.dylib @executable_path/../Resources/lib/libcairo.2.dylib $APP_PREFIX/$target/SpaceAcres.app/Contents/Resources/lib/librsvg-2.2.dylib
```

nazar-pc replied:
I didn't expect for them to be listed explicitly like that. I literally expected something like what I linked above on Windows with a single glob expression. Or maybe something like this (pseudo-code):
```
contents=$APP_PREFIX/$target/SpaceAcres.app/Contents
for path in $BREW_PREFIX/opt/*/lib/*.dylib; do
    cp $path $contents/Resources/lib/$(basename $path)
    install_name_tool -change \
        $(basedir $path) \
        @executable_path/../Resources/lib/$(basename $path) \
        $contents/MacOS/space-acres
done
```

SVG loader will likely remain separate though.

nazar-pc commented on res/macos/run-space-acres.sh:
```diff
@@ -1,17 +1,8 @@
-
-"$LAUNCH_DIR/space-acres"
\ No newline at end of file
+exec "$EXECUTABLE_PATH/space-acres"
```
This makes much more sense, but now I'm wondering if we can get rid of `GDK_PIXBUF_MODULE_FILE`. Can't we know statically where it will be stored or have some kind of installer hook to fix it after extraction?

theluckystrike replied:
@nazar-pc We have a solution is to patch the binary (`libgdk_pixbuf-2.0.0.dylib`), after that we don't need to use `run-space-acres.sh` anymore. What do you think?

nazar-pc replied:
Depends on what kind of patching are we talking about. If something like what we already do already then fine, otherwise I'm okay leaving it as is. The key thing is that we no longer copy to `/tmp` and patch stuff, which is the important thing.

nazar-pc reviewed (commented):
Have you looked into links not working? I suspect something like this is needed on macOS:
https://github.com/subspace/space-acres/blob/3db9689e6646f3fd3b6b3d550fb43c42f7ccfa8b/.github/workflows/release.yml#L179C60-L180

nazar-pc commented:
Tested latest revision, icons still don't render properly

theluckystrike commented:
> Tested latest revision, icons still don't render properly, links don't work either

@nazar-pc Can you test it again (please drag `SpaceAcres.app` to correct place - `/Applications`). I've just tested on a fresh macOS VM and it works fine.

nazar-pc commented:
Already triggered build and asked @DianaPertseva to try it out, thanks!

How do you run macOS VM, anything I can use on Linux with KVM for example?

theluckystrike commented:
@nazar-pc I'm using UTM (https://mac.getutm.app/), it can create macOS VM on a macbook. Few years ago I've tried to use macOS VM on qemu (windows) and kvm (linux): https://github.com/kholia/OSX-KVM, it did work that time but very slow.

nazar-pc commented:
Icons work now, but not links. For example that notification at the top about new version.

theluckystrike commented:
@nazar-pc The open link issue is not package issue but an issue from gtk/glib for macOS: https://gitlab.gnome.org/GNOME/gtk/-/issues/6788 seems gtk team is working on it.

nazar-pc reviewed (approved):
Got it, I think we'll replace LinkButton with a regular button and open the link programmatically ourselves rather than relying on GTK here. Thanks for providing links to upstream issues!

I'll do a few small cleanups and merge, but I think replacing recursive `process_dependencies` with a simpler loop would help with maintenance as well.

Appreciate all the work here and patience with me!

theluckystrike commented:
@nazar-pc I'm testing and plan to commit the changes replacing `process_dependencies` in next hour.

> Appreciate all the work here and patience with me!

My pleasure!
